### ISSUE TITLE
Future-proof serialization

### ISSUE THREAD

**Issue by tdunning** [tdunning/t-digest#87]
Currently, we depend on Java's serialization which is not a good strategy.

It would be much better to do our own and make all the different kinds of Digest use interchangeable formats.

**Comment by Nikko78** [comment]
You can overload Java serialization process to save what you want, for example in my example, methods getBytes and fromBytes already implements in your Digest classes.

In this example, fields of object are ignored and only declared "bytes" field is declared :

public class ExampleSerialize implements Serializable {

    // For serialisation, overload ObjectInput/OutputStream reflection
    private static final ObjectStreamField[] serialPersistentFields = {
        new ObjectStreamField("bytes", byte[].class)
    };
    
        /**
     * Overload {@link ObjectOutputStream#defaultWriteObject()}
     *
     * @param oos object output stream use for write
     * @throws IOException error during write process
     */
    private void writeObject(ObjectOutputStream oos) throws IOException {
        oos.putFields().put("bytes", getBytes());
        oos.writeFields();

    }

    /**
     * Overload {@link ObjectInputStream#defaultReadObject()}
     *
     * @param ois object input stream use for read
     * @throws IOException error during read process
     * @throws ClassNotFoundException class use in serialised stream unknown
     */
    private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
        Object o = ois.readFields().get("bytes", null);
        if (o != null) {
            byte[] b = (byte[]) o;
            fromBytes(this, b);
        }
    }
}

**Comment by anvinjain** [comment]
Isn't it possible to add a Transcoder interface which has `byte[] asBytes(SerializationModel)` and `SerializationModel fromBytes(byte[])` methods? SerializationModel here being a java object that is good enough to represent any digest. TDigest interface can now have a create method that takes SerializationModel as input which all digests implement.

You can provide a default transcoder which does serialization similar to how it's done today (with bugfixes), while keeping the flexibility of accepting PRs from others that implement transcoders for other formats. If this line of thought makes sense to you and more importantly if we can come up with a common model to represent all digests, I will be happy to contribute a PR!

A little background: we have a [project](http://github.com/Flipkart/fk-prof) that does jvm profiling for clusters and all our components talk Protobuf. We deal with latencies, scheduler delays and other metrics and streaming quantiles make a lot of sense. Hence the interest in adopting tdigest but it does not have protobuf support today nor the design is friendly enough to submit a PR.

**Comment by tdunning** [comment]
Are you suggesting that we have different serialization models for, say,
protibuf and any other format?

**Comment by anvinjain** [comment]
No. Basically, I wanted to enable **bring your own serialization logic** with some sane ones provided by the lib. By SerializationModel, I am implying **one** java class something on the lines of:
```java
class SerializationModel {
    int encodingType; double min; double max; float compression; // ... and other members
}

class MergingDigest {
    // ...
    static MergingDigest fromSerializationModel(SerializationModel);
}

interface Transcoder {
    byte[] asBytes(SerializationModel);
    SerializationModel fromBytes(byte[]);
}

class ProtobufTranscoder implements Transcoder {
    // ...
}
```

This keeps the byte magic out of digest implementations and they are responsible for generating a uniform on-heap representation (which is SerializationModel, I know this name sucks)
Users can instantiate any out of the box transcoder or write their own. A way to maintain backward compatibility would be for MergingDigest, etc to optionally take a transcoder as constructor param and use that (or default one if not provided) to keep existing asBytes, fromBytes method working.

If you have any other ideas on how tdigest can support multiple serialization formats (and hopefully you consider that as a feature to be added in the first place), would love to hear about them.

**Comment by tdunning** [comment]
I think your design is fine. It bottlenecks the essential qualities of the
t-digest.

How about we find a better name for the SerializationModel (it might have
other uses, even).  Here are some suggestions that are probably all
problematic, but might get your thoughts going:

DigestEssence
DigestInfo
DigestBasics
NopDigest

I am happy to see a pull request on this.

**Comment by anvinjain** [comment]
Or DigestModel? I see this as a dto representing digests.
Sure, I have some free time next week so expect a PR then.

**Comment by anvinjain** [comment]
You mention interchangeability of digest implementations. Presently, MergingDigest and AvlTreeDigest serialize different values so their serialized representation is not equivalent. How do we come up with a common DTO for both of them?

**Comment by tdunning** [comment]
DigestModel sounds fine!

**Comment by anvinjain** [comment]
@tdunning Need your inputs on what DigestModel looks like. As mentioned in previous comment, presently different fields are serialized across digest implementations

**Comment by tdunning** [comment]
Yes, currently different fields are serialized, but we should synchronize
that.

The minimum set includes:

- compression factor
- min and max values observed
- centroid positions
- centroid weights

Implementations that use point by point insertion may also have special
fields to maintain sort ordering of centroids with the same location (due
to repeated data points, generally). This isn't done yet, but I think it
may get added soon. I am pretty sure that this additional information does
not need to be included in the serialized form if we make the assumption
that centroids are serialized in the internal ordering.

The MergingDigest also stores an extra sizing parameter which determines
the size of the input buffer. This can be dealt with by adding a parameter
to the DigestModel constructor or by just using a default value. Recent
speedups have made this parameter much less important.

Do you see other variations?

**Comment by anvinjain** [comment]
Apart from bufferSize, MergingDigest serializes `lastUsedCell`. For verbose encoding, this is deserialized and set as a property in digest. For small encoding, lastUsedCell additionally represents the values to read and assign in mean/weight array. Rest of the array seems to be zero-filled since `n`(array size) is part of serde as well

**Comment by tdunning** [comment]
In that context, `lastUsedCell` is just the number of active centroids. The
rest of the array is definitely zero filled (and this is even tested, I
believe).

**Comment by tdunning** [comment]
@anvinjain Do you have a PR coming?

We need to cut the new major release. If you can't get this in soon, it will have to wait.

**Comment by anvinjain** [comment]
No, PR will take some time. Have paused work on this for few weeks.
________________________________

**Comment by tdunning** [comment]
OK.

I will leave the current serialization in place but will leave a note that
we may add constructors and methods dealing with this in the future as a
non-breaking API change.

**Comment by tdunning** [comment]
This is being pushed beyond the 3.2 release.

**Comment by anvinjain** [comment]
@tdunning I have raised a PR around this.

**Comment by tdunning** [comment]
Awesome. I will get to this in a few days.

**Comment by anvinjain** [comment]
@tdunning Did you get a chance to look at the PR?

**Comment by tdunning** [comment]
No. I completely spaced looking at it.

Thanks for the ping. Very helpful.

**Comment by anvinjain** [comment]
@tdunning  Ping. Is there a timeline for next release? If you don't see this PR #99 making the cut without major changes, let me know.

**Comment by scheler** [comment]
@tdunning looks like the PR was never merged. This is of interest to us too, so if there are any review comments we can take a look at addressing them too. Thanks in advance.

**Comment by tdunning** [comment]
So ... let's review this again with a lot of experience since last time.

Some thoughts:

- there will be new implementations of digests. It would be nice to be able
to deprecate old formats and use the binary data for new implementations.
What does that imply?

- some digests would like to add special extra data. how does that conflict
with the first point?

- how does this flexibility best play API-wise?

**Comment by amalic** [comment]
Here's my proposal: https://github.com/tdunning/t-digest/pull/162

- Serializations can be configured via `serialization.properties` file.
- The serialization gets instatiated with a `java.util.Properties` object containing all properties form the `serialization.properties` file
- It can load custom implementations via classForName as long as they have been added to the classpath and configured in the `serialization.properties` file

Example for `serialization.properties` file
```
tdigest.serializerClass=org.example.CustomSerialization
someProperty=someValue
someOtherProperty=someOtherValue
```
Hope this is being seen as future proof.

There's still some work to do:
- code review
- proper JavaDoc documentation in source code
- additional tests for custom serialization implementations and property file handling
